### ISSUE TITLE
DragonBones exporter does not export events and Z depth

### ISSUE THREAD

**Issue by davord45** [ndee85/coa_tools#171]
Hi ndee85.

Dont know if you are working on this addon anymore,i see you post a commit from time to time.I gotta report that DragonBones exporter is not finished,if you could please finish it so that it exports Z depth animation and events.I use it in Godot game engine and its awesome,just lacking those features.

Hope you see this.

**Comment by davord45** [comment]
#173 
I made a pull request for this isssue,for anyone having the same problem.

**Comment by ndee85** [comment]
Hi davord45,
do you have a sample file that animates those properties? Do you use a custom build godot version to test?

**Comment by ndee85** [comment]
Implemented zOrder animation with this commit: ceb5a4ce2942204338121e5e18e1c3ddf0613822
Next I am going to implement the event system.

**Comment by davord45** [comment]
Sorry for getting back so late,i am using Godot Dragonbones module which you can download here:

https://www.godotdragonbones.com/

**Comment by davord45** [comment]
I can make you a sample file tommorow.

**Comment by davord45** [comment]
Thanks for taking the time to implement these features btw!

**Comment by davord45** [comment]
Also the events i managed to export only Sound and Play type not regular events cause they are a Array of events and the JSON formatter crashed. Hopefully you can figure that out. Right now in my pull request In COA regular events are play events.

**Comment by ndee85** [comment]
I have checked the regular events yesterday. I think it shouldn't be to hard to add. I will have a look at it when I have some spare time.

**Comment by ndee85** [comment]
I didn't know that extra godotdragonbones version existed :-)
Good to know.

**Comment by davord45** [comment]
I tested out zOrder export and it works great!
Way more elegant than what i had lol.

**Comment by ndee85** [comment]
Great. Glad to hear!
Goint to add the event system soon and than I think the exporter should be more or less complete. Or do I still miss something?

**Comment by davord45** [comment]
Well i found that meshes that are Slots which have a ShapeKey driver deformation, those deformations dont get exported like regular Mesh ShapeKey deformations. And animation curves always get exported as a standard S curve. Also i found that when the bone scale for instance X goes to negative value it exports Y value as a negative as well,even when its positive. I have implemented some fixes for that. Do you think it is possible also to make 2 bone IKs to export as DragonBones IK? Cause i know that IKs get baked into chain bones.

**Comment by davord45** [comment]
Slots deformation and scale are kind of a bug and Curve and IK are more of a new feature.

**Comment by ndee85** [comment]
the problem with ik is that they don't behave the exact same way as in dragonbones. Blender is more powerful in that regard. So I thought the best way would be to bake those. Based on the setup it can differ quite a lot in blender. So think the best way is to bake in that case. It is similar with the curves. Blenders graph editor is way more flexible, end exporting this is not as easy. :-)

**Comment by davord45** [comment]
Understandable,i managed to export curves but in some cases when the curve is really extreme DragonBones runtime starts freaking out.

**Comment by ndee85** [comment]
Ok. Timeline Events are now fully supported. Please have a test with your files and see if they work as expected!

**Comment by ndee85** [comment]
Forgot to commit some changes. So now everything should be updated.
ea3dc7e6a2bf616d0e726fdd07f3c8ed4192fbf3

**Comment by davord45** [comment]
I added the events but they dont export. And now i got a bug with the zOrder as well,i made a simple one arm go behind the body and other go in front.It made the keyframes but the only plays second arm going in front of the body but on a keyframe where first arm goes behind the body.

**Comment by davord45** [comment]
Do you want me to send you the .blend to test with it?

**Comment by davord45** [comment]
Here is the link to the .blend file:
https://www.mediafire.com/file/e02akkpnvz2ov5m/nDeechaser.blend/file

**Comment by ndee85** [comment]
Thanks. I'll have a check with your file an see what is causing the problem.

**Comment by ndee85** [comment]
Are you sure you updated the sources correctly? I just tested your file and all events did export properly. Also adding multiple sound events should not be possible on the same frame. Do you have an "Animation" event? If not, you did not update correctly.

Just checked that drawing order is still not working correctly in some cases. Will have a recheck on that.

**Comment by davord45** [comment]
I didnt have "Animation" type,I downloaded the commit from blender28 and installed it to Blender. Do i need to update in Blender as well?

**Comment by ndee85** [comment]
make sure to zip the coa_tools directory from Blender directory. Install that zip file then.

**Comment by davord45** [comment]
You are right i was downloading the wrong branch. Now exporting works very nicely, I like how you give on Animation events dropdown of all animations! May i ask how do Target,Int,Float,String fields on regular events effect JSON and what is their use in DragonBones?

**Comment by ndee85** [comment]
I have to admit  I am not sure how to use the custom event. I asume it will trigger a callback in the runtime with the given parameters that can be set. The rest will be up to the dev what callback will be doing. But those are just assumptions.

I just reverse engineered the json file that dragonbones exports and made sure it outputs the same structure.

**Comment by ndee85** [comment]
Just checked the godotdragonbones. It seems there are two signals in godot. One is for triggering sounds and the other is used to trigger customEvents.

dragon_anim_snd_event
dragon_anim_event

The dragon_anim_event will have a dictionary parameter which delivers the int, float and string props.

**Comment by davord45** [comment]
Nice, that does seem useful actually.

Was wondering could you add a field on the export panel for naming the armature? Cause right now everything that gets exported is called "Armature". Godot Dragonbones module has a way to control child armatures that are parented to a bone of another armature.

**Comment by ndee85** [comment]
ok. Just added the armature name in the export settings.
You can also update COATools via the addon preferences. There for open the addon settings.
Refresh and then install the blender28 branch. Just force the install. This will get the latest source from the github repo.

**Comment by davord45** [comment]
I get this error when i try to export. But i dont have slots in this armature.
![image](https://user-images.githubusercontent.com/70846358/102254358-cd34a600-3f08-11eb-879a-d44a9ecfd133.png)

**Comment by ndee85** [comment]
Is this the exact same file you posted above? Strange exports fine for me.

**Comment by ndee85** [comment]
Also just added a check which makes sure that slot_data does exists before accessing it. Not sure if this will fix your issue. Or something else is the root. But please make sure to restart blender after updating to latest branch.

**Comment by davord45** [comment]
It was the same. I updated it but i dont see a check for slots. But it worked now correctly.

**Comment by ndee85** [comment]
Ok. Still working on exporting the zOrder correctly. I noticed some problems which I need to think of how to best fix.

**Comment by davord45** [comment]
If it helps i noticed that the zOrder in DragonBones is looking at the default order and then comparing to the changed order. What is the fastest way of rearranging the order of meshes. It does not compare changed order to the previous order in animation,just the default one.
